### DISCUSSION TITLE
NestJS server auth

### DISCUSSION THREAD

**Discussion by Yonben** [pocketbase/pocketbase#1934] [Q&A] [answered]
Hi,
I looked around but might be missing something.
I'd like to do some additional operations on my backend, build with Nest, from user requests.
Therefore I need a way to validate my user is auth properly.

Is there any example of anyone implementing something like that?

Thanks!

**Comment by ganigeorgiev** [comment, marked as answer]
I'm not sure that I understand the question.

You can find JS examples how to authenticate as admin or app user in https://pocketbase.io/docs/authentication/.

If the above doesn't help, please provide a little more context what you are trying to do and/or what is the expected user flow (pseudo-code is also ok).

**Reply by Yonben** [reply]
Hi,
Sorry for the lack of clarity.
I have a frontend client, it's authed with Pocketbase with the js sdk.
I also have a NestJS backend. It exposes an api to the client by I'd like my api calls to be authenticated, so the actions the rest api does - except admin actions - are done with the user auth.

Is it clearer?

**Reply by ganigeorgiev** [reply]
The described use case is too abstract/generic and I'm still not sure that I understand what you are trying to do.

PocketBase doesn't have sessions, and a single request is considered authenticated as long as it has a valid `Authorization: TOKEN` header.

When you authenticate via the JS SDK in the browser, the token is stored by the `pb.authStore` in `localStorage`. In node env the `authStore` data persists for the duration of the SDK instance.

If I understand your use case correctly, you want to make requests against the PocketBase APIs from both browser (`browser->PocketBase`) and node (`browser->node->PocketBase`), right?

One possible approach could be to pass the `pb.authStore.token` to your node action calls and create a one-off PocketBase SDK client instance with it, for example:

```js
@Get()
yourAction(@Req() request: Request): string {
  const token = request.get("Authorization");

  const pb = new PocketBase("http://127.0.0.1:8090");

  pb.authStore.save(token, null);

  // call authRefresh() if you want to validate the token
  // and to get the latest user data in pb.authStore.model
  // await pb.collection("users").authRefresh();

  // this will send a GET /api/collections/example/records?... request with Authorization: ${token}
  const result = await pb.collection("example").getList(1, 30);

  return ...;
}
```

If you have several similar node endpoints, you may want to move the above in a middleware, create a custom decorator, etc.

---

Another approach if you want to be able to modify the user auth state from both browser and node seamlessly (eg. to "login" or "logout" a user from a node action) could be to use a cookie-based flow. The default `pb.authStore` comes with 2 helper methods for working with cookies:

```js
// update the store with the parsed data from the cookie string
pb.authStore.loadFromCookie('pb_auth=...');

// exports the store data as cookie string, with option to extend the default SameSite, Secure, HttpOnly, Path and Expires attributes
pb.authStore.exportToCookie({ httpOnly: false }); // Output: 'pb_auth=...'
```
For this scenario, you'll have to maintain 2 separate SDK instances and to export the cookie with `{httpOnly: false}`:

1. Browser client:
```js
import PocketBase from "pocketbase";

const pb = new PocketBase("http://127.0.0.1:8090");

pb.authStore.loadFromCookie(document.cookie)

pb.authStore.onChange(() => {
  document.cookie = pb.authStore.exportToCookie({ httpOnly: false });
});
```

2. Node client
  _(this is similar to the above with the difference that the cookie is loaded from the req object and updated via the res object; the client must be one-off, aka. live only for the duration of the node request-response)_:

```js
import PocketBase from "pocketbase";

const pb = new PocketBase("http://127.0.0.1:8090");

pb.authStore.loadFromCookie(req.headers.cookie || '');

pb.authStore.onChange(() => {
  res.setHeader('Set-Cookie', pb.authStore.exportToCookie({ httpOnly: false }));
});

// call authRefresh() if you want to validate the loaded cookie
// inside the node server and to get the latest user data in pb.authStore.model
// await pb.collection("users").authRefresh();
```

**Reply by Yonben** [reply]
Ok that's exactly my use case.
I thought about using the exportToCookie, but I thought only the server could load/add a HttpOnly Secure cookie, not the client itself.
If localStorage or not httpOnly cookie, then it's prone to be stolen/manipulated afaik no?

My other worry was the "one-off" pb instances. Is there any downside to short lived instances? If not then great :)

Thanks a lot for the detailed answer!

Edit to add:
After re reading I see I have to use `httpOnly: false`, which I find off putting indeed.
So the solution would be to add an Authorization header with the token. I missed the `save` function in my searches.
And for everything in the browser I can uses the authStore to login and exportToCookie with httpOnly I guess.
The question remain for the one-off instances impact on the server side.

Thanks again I'll try that :)

**Reply by ganigeorgiev** [reply]
> If localStorage or not httpOnly cookie, then it's prone to be stolen/manipulated afaik no?

Only if your application is vulnerable to XSS, but for that you can always define a CSP (you can also check https://github.com/pocketbase/js-sdk#security).

> The question remain for the one-off instances impact on the server side.

It doesn't have impact on the performance if that is your concern. The SDK memory footprint is also minimal and you can think of it like just creating a new local object. You need to maintain a separate client instance for each node request-response because otherwise you risk requests from different users to conflict/overwrite the same `pb.authStore` state. This is not an issue in the browser because the client/user is only 1.

**Reply by Yonben** [reply]
Great, thank you very much for this detailed and thorough answer 🙏
